File: Leetcode0450.java
package com.dezhonger.y2020.m01.d25;

import com.dezhonger.y2019.TreeNode;

/**
 * Created by dezhonger on 2020/01/25
 *
 * @author dezhonger
 * @since 2020/01/25
 */
public class Leetcode0450 {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;
        if (key > root.val) root.right = deleteNode(root.right, key);
        else if (key < root.val) root.left = deleteNode(root.left, key);
        else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;
            TreeNode tmp = root.left;
            while (tmp.right != null) tmp = tmp.right;
            tmp.right = root.right;
            return root.left;
        }
        return root;
    }

}

File: Leetcode0450.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        del(root, key);
        return root;
    }
    void del(TreeNode* &root, int key) {
        if (root == NULL) return ;
        if (key == root -> val) {
            if (!root -> left && !root -> right) root = NULL;
            else if (!root -> left) root = root -> right;
            else if (!root -> right) root = root -> left;
            else {
                //找到后继节点
                TreeNode* p = root -> right;
                while (p -> left) p = p -> left;
                root -> val = p -> val;
                del(root -> right, root -> val);
            }
        }
        else if (key < root -> val) del(root -> left, key);
        else del(root -> right, key);
    }
};
